tests/testing_common.py

```diff
@@ -300,8 +307,9 @@ def _test_merge_layers(self, model_id, config_cls, config_kwargs):
 
             logits_transformers = transformers_model(**dummy_input)[0]
 
-            self.assertTrue(torch.allclose(logits_lora, logits_merged, atol=1e-4, rtol=1e-4))
-            self.assertFalse(torch.allclose(logits_merged, logits_transformers, atol=1e-10, rtol=1e-10))
+            self.assertTrue(torch.allclose(logits_unmerged, logits_merged, atol=1e-4, rtol=1e-4))
+            if config_cls == LoraConfig:  # merge does not change logits for IA3
```

Comment:
can you elaborate on that? what do you mean specifcally by `merge does not change logits for IA3`? if merging is not supported I think we should just raise an error. WDYT?

Reply:
So merging is supported, but the specific check here is testing for a difference between `logits_merged`  (after merging IA3 vectors) and `logits_transformers` (without IA3 vectors). With IA3, the initialization is all ones initially (From the paper: "...[the learned vectors] are all initialized with ones so that the overall function computed by the model does not change when they are added." In the case of LoRA, the random initialization will change model outputs ever so slightly after adding LoRA weights, which is being tested with the `assertFalse` statement.  This isn't going to be true for IA3 -  so I have simply ignored that test. I'm not sure what other tests can be added though for merging.

Reply:
I see thanks a lot for the detailed explanation ! Do you think it is possible to force the initialization to be something different than all ones.? I think in the past I have faced the same issue with LoRA because the B matrix was initialized to all zeros. I have added a boolean here: https://github.com/huggingface/peft/blob/06fd06a4d2e8ed8c3a253c67d9c3cb23e0f497ad/src/peft/tuners/lora.py#L91 let me know if you need help on that ! 
 Otherwise all good I think, it can be done in a follow up PR

Reply:
Ah okay! There's actaully an argument already called `init_ia3_weights` that's supposed to give this control - but even with `init_ia3_weights` as False, currently the initialization is all ones : 
https://github.com/SumanthRH/peft/blob/5ed92fad11aeec3ec94c8dfd908e81a3cb09d4e4/src/peft/tuners/ia3.py#L399

If this going to be different, then the initialization should be such that the merged version is close enough to the pre-trained weights, which means that the entries in the learned vectors should be close enough to one (because they rescale activations). I'm not too sure what standard initialization fits in here. Maybe something like PyTorch's nn.Linear [initialization](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) for matrices, but shifted by one: $U(-\sqrt{k}, \sqrt{k}) + 1$ , where $k=1/\text{vector-length}$.  I also don't want to make it look like we conjured up an initialization, so I'm open to hearing your thoughts on this!

Reply:
@younesbelkada Could you explain why we want to require the outputs to be different each time when initializing the model? Is it not actually a good thing if by default, the added weights lead to an identity operation? I'm running into the same test error when using LoRA with certain custom models.

Reply:
Hello @BenjaminBossan, LoRA initializes `lora_B` to all 0 so that we start with an identity function. So, for all usage `init_lora_weights=True`. Only during testing when we can't actually train the model but want to test the functionality of `merge_and_unload`, `init_lora_weights` is set to False else even after merging it will be identity op and we can't test if the merge was successful or not.

Reply:
Ah I see, thanks for explaining. This should probably added as a comment to that line, since I think I'm not the only one who was confused.

Regarding the testing of the merging feature, would it be possible to check the weights directly, instead of the model outputs?